### ISSUE TITLE
Add automatic validation feature to dataset (data.yaml and other aspects of dataset structure)

### ISSUE THREAD

**Issue by Gabrli** [ultralytics/ultralytics#21492]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I propose integrating a built-in validation step before training that checks the integrity of the `data.yaml` file and related dataset configuration.

This would help users catch common mistakes early (e.g., wrong number of classes, incorrect paths, invalid annotation formats).

### Functionality

✅ Check if nc in data.yaml matches the number of names and the max class index in label files

✅ Validate that paths listed in train/val keys exist and contain supported image formats

✅ Ensure all YOLO annotation .txt files contain 5 values per line (class, x, y, w, h)

✅ Detect empty or malformed label files

✅ Optional: Warn or fix mismatched image sizes 

✅ Return summary of all checks, e.g., "3 label files malformed", "2 images too small", etc.

**Could be exposed via CLI like:**

`yolo check dataset --structure --annotations --sizes --expected-size 640 480`

---

Many users (especially beginners) experience training failures due to simple misconfigurations such as:

- Wrong nc not matching actual labels
- Invalid or missing annotation files
- Incorrect folder structures (e.g., missing images/val)
- Unexpected image sizes or label class indices

Currently, the model crashes mid-training, which wastes time. Having an integrated check would prevent this by validating the dataset before training begins.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
There's already label validation that's performed before training:

https://github.com/ultralytics/ultralytics/blob/b905114f46311976c816d586073579a1a0b29266/ultralytics/data/utils.py#L180

**Comment by Gabrli** [comment]
> There's already label validation that's performed before training:
> 
> [ultralytics/ultralytics/data/utils.py](https://github.com/ultralytics/ultralytics/blob/b905114f46311976c816d586073579a1a0b29266/ultralytics/data/utils.py#L180)
> 
> Line 180 in [b905114](/ultralytics/ultralytics/commit/b905114f46311976c816d586073579a1a0b29266)
> 
>  def verify_image_label(args: Tuple) -> List:

Ok, good know that. Btw sorry for didn't check this. Do you think, rest of functionality which IS described would be useful?

**Comment by Y-T-G** [comment]
I think most of the highlighted checks are present:

- `nc` check with actual labels is present 
- Invalid annotation checks are present
- Missing images checks are present
- Unexpected image size checks are not present. But that's because any image size works with Ultralytics. So there's no such thing as an unexpected image size.

**Comment by Gabrli** [comment]
Thanks a lot for the clarification — that makes sense.

I totally get that some checks already happen before training (like missing labels or invalid indices), which is great. What I’m proposing is to take it one step further and give users a clearer, more explicit "preflight check" before training kicks off — not just to catch issues, but to explain them clearly.

For example:

- If a .txt label file is empty — tell the user: “Label file X is empty”
- If a line has only 4 values instead of 5 — show: “Malformed annotation in file Y, line Z”
- If nc in data.yaml is 3, but you’ve got a label with class 5 — surface that mismatch
- If train: points to a non-existent or empty folder — highlight it directly

Or even just a final message: “✅ 417 images, 415 valid labels, 2 empty files, 1 malformed line”

The goal isn’t to replace anything — more to wrap what’s already there in a clear summary that helps especially beginners (who might otherwise just see an error halfway through training and not know why it failed).

I was thinking of exposing it as something like:

`yolo check dataset --annotations --structure`

Appreciate your feedback either way :)

**Comment by glenn-jocher** [comment]
Thanks for the detailed clarification! I understand you're proposing a user-friendly wrapper around the existing validation with clearer messaging and a dedicated CLI interface - that's actually a great idea for improving the user experience.

You're right that while we have robust validation internally (like `verify_image_label` and the checks in `check_det_dataset`), exposing this as an explicit pre-training check with clear summaries would be very helpful, especially for beginners who currently only see errors mid-training.

A CLI command like `yolo check dataset` that provides a comprehensive validation report with clear messages about empty files, malformed annotations, and dataset statistics would definitely add value. Feel free to submit a PR - this would be a nice enhancement to make the existing validation more accessible and informative for users!

**Comment by Gabrli** [comment]
Thanks for your reply. I'm really happy to be allow to provide new functionality.

I was thinking about this functions and maybe nice one would be to automatic fix function to correct detected issue?

For example if nc dosen't match to names length - CLI command is asking "update `nc` to current `names` length? (yes/no):" and then software can change number in nc to current names length. Also with the anothers errors. 

Thank you for your time!

**Comment by glenn-jocher** [comment]
That's a thoughtful extension to the validation feature! Auto-fix functionality could be really helpful for common issues like `nc`/`names` mismatches or empty label file cleanup.

I'd suggest starting with the safer fixes first (like updating `nc` to match `names` length) and being more cautious with structural changes that could affect training data. Interactive prompts asking for user confirmation before making changes would be a good approach to keep users in control.

Feel free to include this in your PR - you could implement it as an optional `--fix` flag alongside the validation command.

**Comment by Gabrli** [comment]
Ok, thanks for your time! :)

**Comment by glenn-jocher** [comment]
Great! Looking forward to your PR. When you start working on it, the existing validation functions in `ultralytics/data/utils.py` like `verify_image_label` and `check_det_dataset` will be good reference points for the underlying checks you'll want to surface in the CLI interface.

**Comment by Gabrli** [comment]
Hi, I'm finishing my work on this function, so you can expected PR soon, but can you ensure me, if I want to add new CLI command I have to do this in [main/ultralytics/cfg/__init__.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/__init__.py) in `entrypoint` function?

**Comment by glenn-jocher** [comment]
Yes, that's correct! You'll want to add your new CLI command to the `special` dictionary in the `entrypoint` function in `ultralytics/cfg/__init__.py`. 

You can see the existing special commands like `"checks"`, `"hub"`, `"settings"` etc. are defined there. For your dataset validation feature, you'd add something like `"check": lambda: handle_dataset_check(args[1:])` to that dictionary, then implement the `handle_dataset_check` function to parse the dataset-specific arguments and call your validation logic.

**Comment by Gabrli** [comment]
Ok, thank you so much :)

**Comment by glenn-jocher** [comment]
You're welcome! Excited to see your PR when it's ready. The dataset validation feature sounds like it'll be a great addition for the community.

**Comment by Gabrli** [comment]
@glenn-jocher Hi, If I want to try write unit tests for my new feature and I'm thinking where I'm supposed to do that? In [here](https://github.com/ultralytics/ultralytics/blob/main/tests/test_cli.py) ? Also I have question about docs. I have to add new MD file for example: `data_validation.md` or add docs to [cli.md](https://github.com/ultralytics/ultralytics/blob/main/docs/en/usage/cli.md) ?

**Comment by glenn-jocher** [comment]
For unit tests, `tests/test_cli.py` is the right place since you're adding a CLI command. You might also want to add specific validation tests in a separate test file if your feature has substantial logic beyond the CLI interface.

For documentation, add it to the existing `cli.md` file since it's a CLI command - that's where all the other `yolo` commands are documented. You can add a new section for the `yolo check` command alongside the existing train, predict, val, etc. sections.

**Comment by Gabrli** [comment]
Hi, I've pushed [PR](https://github.com/ultralytics/ultralytics/pull/21599) to reviewed. Also thank you for your time!

**Comment by glenn-jocher** [comment]
Perfect! Thanks for submitting the PR - I'll take a look at it soon. Great work on implementing the dataset validation feature with CLI integration. The community will definitely benefit from having clearer dataset validation feedback before training begins.

**Comment by Gabrli** [comment]
Hi, when I could expect first review for my PR? 

Thanks for your time!

**Comment by glenn-jocher** [comment]
Thanks for the PR—no ETAs on reviews, but we’ll look as soon as we can; in the meantime, please keep the changes scoped, ensure CI is green, and include tests/docs per the [Contributing Guide](https://docs.ultralytics.com/help/contributing/).

**Comment by Gabrli** [comment]
Ok, thank you so much for your help!

**Comment by glenn-jocher** [comment]
You're welcome—thanks for contributing; we’ll follow up on your PR with any feedback.
